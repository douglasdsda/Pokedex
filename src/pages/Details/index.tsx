import React, { useState, useEffect, useCallback } from 'react';

import { FiArrowLeft } from 'react-icons/fi';
import { Link, useRouteMatch } from 'react-router-dom';
import {
  Container,
  Main,
  Footer,
  FooterContent,
  Back,
  ListTree,
} from './styles';
import Header from '../../components/Header';

import CardInfo from '../../components/CardInfo';
import { usePokemon, Pokemon, propsTypeIn } from '../../hooks/pokemons';
import api from '../../services/api';
import { formatUpperCase } from '../../utils/utilsHelper';

interface DetailsParams {
  pokemon: string;
}

const Details: React.FC = () => {
  const { setPokemon } = usePokemon();

  const { params } = useRouteMatch<DetailsParams>();

  const [pokemonDetail, setPokemonDetail] = useState<Pokemon>();
  const [tree, setTree] = useState<Pokemon[]>([]);

  const getNamesEvelutions = useCallback(async (data: any, myName: string) => {
    let evolution;

    evolution = data;

    const names: string[] = [];
    while (evolution) {
      names.push(evolution.species.name);
      [evolution] = evolution.evolves_to;
    }

    return names.filter(name => name.toUpperCase() !== myName.toUpperCase());
  }, []);

  const setTreePokemons = useCallback(
    async (names: string[], list: Pokemon[]) => {
      for (let index = 0; index < names.length; index += 1) {
        const family = await api.get(`pokemon/${names[index]}`);
        const name = formatUpperCase(family.data.name);
        const formattedTypes = family.data.types.map((item: propsTypeIn) => {
          return { ...item, name: formatUpperCase(item.type.name) };
        });

        const item = {
          id: family.data.id,
          name,
          sprite: family.data.sprites.other.dream_world.front_default,
          idPokemon: family.data.id,
          types: formattedTypes,
          stats: family.data.stats,
          weight: family.data.weight,
          height: family.data.height,
        };

        list.push(item);
      }
    },
    [],
  );

  useEffect(() => {
    const load = async () => {
      const data = await setPokemon(Number(params.pokemon));

      setPokemonDetail(data);

      const species = await api.get(`pokemon-species/${data.idPokemon}/`);
      const evulutions = await api.get(species.data.evolution_chain.url);

      const names = await getNamesEvelutions(evulutions.data.chain, data.name);
      const list: Pokemon[] = [];
      await setTreePokemons(names, list);
      setTree([...list]);
    };

    load();
  }, [setPokemon, params.pokemon, getNamesEvelutions, setTreePokemons]);

  return (
    <Container>
      <Header />
      <Main>
        <Back>
          <Link to="/">
            <FiArrowLeft />
            <span>Back</span>
          </Link>
        </Back>

        {pokemonDetail && (
          <CardInfo
            data-testid="Item-Card"
            containerStyle={{ margin: '0 auto' }}
            key={pokemonDetail.id}
            idPokemon={pokemonDetail.idPokemon}
            num={pokemonDetail.id}
            name={pokemonDetail.name}
            sprite={pokemonDetail.sprite}
            types={pokemonDetail.types}
            weight={pokemonDetail.weight}
            height={pokemonDetail.height}
            phys
            status
            stats={pokemonDetail.stats}
          />
        )}
      </Main>
      <Footer>
        <h2>Family Tree</h2>
        <FooterContent />
        <ListTree>
          {tree.map((item: Pokemon) => {
            return (
              <CardInfo
                key={item.id}
                idPokemon={item.idPokemon}
                num={item.id}
                name={item.name}
                sprite={item.sprite}
                types={item.types}
                titleAndSubTitle
              />
            );
          })}
        </ListTree>
      </Footer>
    </Container>
  );
};

export default Details;
