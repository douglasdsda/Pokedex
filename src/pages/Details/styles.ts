import styled from 'styled-components';

export const Container = styled.div``;

export const Main = styled.div`
  padding: 8px 30px;
  display: flex;
  flex-direction: row;

  div {
    svg {
      color: #ff9000;
      height: 24px;
      width: 24px;
    }
  }
`;

export const Card = styled.div`
  display: flex;

  flex-direction: row;
  align-items: flex-start;

  height: 240px;
  width: 240px;

  margin: 0 auto;

  background: #3e3b47;
  border-radius: 8px;
  padding: 10px;
  cursor: pointer;
`;

export const CardHeader = styled.div`
  display: flex;

  flex-direction: column;
  align-items: center;

  flex: 1;

  img {
    margin: 0 auto;
    height: 110px;
    width: 110px;
  }
`;
