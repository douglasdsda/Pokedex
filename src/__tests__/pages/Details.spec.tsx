import React from 'react';
import { render } from '@testing-library/react';

import Details from '../../pages/Details';

const mockedHistoryPush = jest.fn();

jest.mock('react-router-dom', () => {
  return {
    useHistory: () => ({
      push: mockedHistoryPush,
    }),
    useRouteMatch: () => ({ params: 'id' }),
    Link: ({ children }: { children: React.ReactNode }) => children,
  };
});

jest.mock('../../hooks/pokemons', () => {
  return {
    usePokemon: () => ({
      setPokemon: () => {
        return {
          idPokemon: 1,
        };
      },
    }),
  };
});

describe('Details Page', () => {
  it('should be able to datails in', async () => {
    const { debug } = render(<Details />);
    debug();
  });
});
