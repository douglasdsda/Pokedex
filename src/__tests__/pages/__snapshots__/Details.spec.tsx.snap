// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Page should be able to datails in 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm StDqN"
  >
    <div
      class="sc-AxheI nqLDC"
    >
      <img
        alt="Pokébola"
        src="pokebola.svg"
      />
      <h1>
        POKÉMON CHALLENGE
      </h1>
    </div>
    <div
      class="sc-AxirZ kowbRK"
    >
      <div
        class="sc-AxhUy jlfIlr"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="19"
            x2="5"
            y1="12"
            y2="12"
          />
          <polyline
            points="12 19 5 12 12 5"
          />
        </svg>
        <span>
          Back
        </span>
      </div>
    </div>
    <div
      class="sc-AxiKw kMnbKe"
    >
      <h2>
        Family Tree
      </h2>
      <div
        class="sc-AxhCb knfRmQ"
      />
      <div
        class="sc-AxgMl eqmcpW"
      />
    </div>
  </div>
</DocumentFragment>
`;
